### ISSUE TITLE
Support for expanded 'operators' (e.g. In) in field selectors

### ISSUE THREAD

**Issue by munnerz** [kubernetes/kubernetes#128154]
### What would you like to be added?

Support for selecting objects using 'set operations' (notably 'In') with field selectors, bringing parity with labelSelectors.

More specifically, I'd like to be able to select over the `metadata.namespace` field, so I can establish multi-namespace watches with consistent RV semantics.

### Why is this needed?

To perform a multi-namespace list/watch today, a user must fetch objects at the cluster-scope and then filter down the response client-side. This is expensive and doesn't scale well as the size of the cluster grows (and various permissions/authZ boundaries are put in place).

Forcing the apiserver to perform this sort of filtering has typically been avoided in the past, as etcd does not natively support something like this, meaning apiservers have to establish a cluster-scoped watch for each client and perform filtering in memory, leading to high amounts of decoding objects from etcd (only to throw them away again after).

Recent work within the apiserver I believe enables us to reduce the impact of these kinds of requests, namely:

* https://github.com/kubernetes/enhancements/issues/2340 (consistent reads from cache)
* https://github.com/kubernetes/enhancements/issues/4568 (resilient watch cache initialization)

Combined, these allow us to ensure we serve list/field selectors from the watch cache, allowing us to optimise and index (similar to how we index certain field selectors and label selectors today).

I believe this feature would complement https://github.com/kubernetes/enhancements/issues/4601 (authorise with selectors) well, as it'd allow end-users to authorise field selector usage as well.

---

For this feature, I think it'll be important to ensure we also index watches and probably also list operations that utilise fieldSelectors, at the very least for `metadata.namespace`. For watches, this is relatively simple as we can add the `cacheWatcher` for the request to the `indexedWatchers` structure multiple times (by generating multiple namespaceName/scope entries): https://github.com/kubernetes/kubernetes/blob/06a15c5cf96131faaf44f93f1be228a013ae5c0d/staging/src/k8s.io/apiserver/pkg/storage/cacher/cacher.go#L135-L163. I've hacked together a quick PoC of the sort of change needed for watches here: https://github.com/munnerz/kubernetes/commit/ae075a0d30906a7218fdc123dd11cc96748bc20d - please do note that this relies on stuffing a `kubernetes.io/metadata.namespace` label into GetAttr funcs to emulate fieldSelector 'In' support, but it demonstrates the performance optimisation technique somewhat :)

Optimising LIST operations is a bit more complex today (as far as I can tell!) due the underlying Store/Indexer implementations in `k8s.io/client-go/tools/cache` not supporting indexed 'set operations' natively.

**Comment by sftim** [comment]
I agree with all this, but the `notin` operator is perhaps even more important in one respect: you can only emulate it client side by doing a full list (or watch) and then iterating through the collection. For `in`, you can at least emulate it somewhat using multiple watches.

**Comment by jpbetz** [comment]
Given the recently added "CRD field selector" and "authorize with selectors" enhancements, this is a great good time to discuss this.

I'm generally supportive of bringing field selectors in parity with label selectors.

Do we need to tightly couple field selectors operator support with indexing support?  I see value to both, but it's not clear to me that it's better to couple them together vs. introducing them as independent changes and indexing brings up resource utilization considerations.

**Comment by liggitt** [comment]
Values for label selectors are pretty tightly bounded (length, charset, etc)... values for field selectors are not at all, so the inputs to valid field selectors can be more expensive than label selectors

We would also have to figure out how to evolve field selector parsing to handle `in/notin (...)` syntax, and how new clients / old servers would interact (hopefully old servers would just error sensibly if presented with a field selector they cannot parse). I think the parsing change would primarily mean escaping `(` and `)` characters, but only in values inside an `in/notin` expression - xref https://github.com/kubernetes/kubernetes/pull/28112 for prior hoops we had to jump through when adjusting field selector parsing / escaping

**Comment by munnerz** [comment]
> Do we need to tightly couple field selectors operator support with indexing support? I see value to both, but it's not clear to me that it's better to couple them together vs. introducing them as independent changes... indexing brings up resource utilization considerations.

I don't think we _have_ to, however not indexing also has its own resource utilisation and performance considerations. Though I suppose given all users that can use a label selector at the cluster scope, can also take action at the cluster-scope, this isn't a net-change in the 'attack surface' for overloading an apiserver for someone with equivalent permissions today, so splitting the two may make sense.

> Values for label selectors are pretty tightly bounded (length, charset, etc)... values for field selectors are not at all, so the inputs to valid field selectors can be more expensive than label selectors

When you say expensive here, do you mean to parse the selector in the first place? Or because they can be larger in size? (both keys and values).

> hopefully old servers would just error sensibly if presented with a field selector they cannot parse

🤞 we will need to look into how an apiserver behaves if we submit something like `mything in (a, b=c, d)` - does it reject it, or does it interpret it as `mything in (a, b` = `c, d)`? (and variations thereof..). We'd need a multi-step release to tighten this validation otherwise (once we have asserted that all existing known fieldSelector keys would not be excluded with this new policy).

**Comment by liggitt** [comment]
> We'd need a multi-step release to tighten this validation otherwise (once we have asserted that all existing known fieldSelector keys would not be excluded with this new policy).

Looks like current apiserver doesn't complain about field selectors using in/notin syntax as long as there's an `=` or `!=` somewhere in each comma-split term. It happily accepts all of the following :-/

```
		`a in (b=c)`,
		`a in (b\=c)`,
		`a in (b=c,d=e)`,
		`a in (b=c,d\=e)`,
		`a in (b\=c,d=e)`,
		`a in (b\=c,d\=e)`,
		`a notin (b=c)`,
		`a notin (b\=c)`,
		`a notin (b=c,d=e)`,
		`a notin (b=c,d\=e)`,
		`a notin (b\=c,d=e)`,
		`a notin (b\=c,d\=e)`,
```

My first thought is that we'd want to constrain the allowed values for `in`/`notin` expressions to ones old servers would error on, but I haven't thought deeply about it.

**Comment by munnerz** [comment]
Another alternative could be to use a prefix to denote a schema version for the fieldSelector, using something that'd cause old apiservers to return an error. Something like: `?fieldSelector=v2,something in (a, b, c)`. This is assuming that the apiserver will balk at seeing a string without a `=` separator 'contained' by a comma.

I'm not keen on inventing adhoc versioning schemes, and this scheme may not be applicable/useable for *other* query parameters. I feel it'd require a lot of discussion/thought/consideration in sig-arch to make such a change.

**Comment by jpbetz** [comment]
Crazy idea. Tell me if you hate it.

What if we kept label and field selectors as-is introduced a new `selector` field that we can define a grammar for without being constrained by grammar of field selectors?  We could support both field and label selectors in it. For example: `field.type in (a, b) && label.env notin (prod)`?

Is it worth hitting the reset button on the grammar?  We could probably do a lot better knowing what we know now (It wouldn't necessarily need to be the CEL grammar, but maybe aligning with CEL at least grammatically wouldn't be terrible either, even if we only support a very small subset of the operators)?

**Comment by munnerz** [comment]
I guess one issue with a new field/parameter is again how old servers will behave. They don't perform strict decoding of query parameters meaning no selection criteria will be applied. For deletecollection, this would be pretty catastrophic.

I think to introduce a new DSL, we need to find a way to break parsing in all existing servers. That could be with a string with no '=' being provided in the fieldSelector field by new clients (and then defining a new field name to store the new schema selector), or could otherwise be stuffed into the existing field after the prefix (if we think it's safe to do so & preferable).

**Comment by liggitt** [comment]
> Crazy idea. Tell me if you hate it.

I like the idea of finding a way to get the grammar we want without jumping through insane hoops to make sure it doesn't get accidentally accepted by old servers

I don't love that old servers would just ignore the `selector` query param, so the failure mode of sending a new filter to an old server is that it returns stuff unfiltered.

> Another alternative could be to use a prefix to denote a schema version for the fieldSelector, using something that'd cause old apiservers to return an error. Something like: ?fieldSelector=v2,...

Something like that is still slightly hacky but might get us the properties we want. Would maybe want to define the normalized serialization so that a field selector that doesn't make use of in/notin with multiple values would still serialize in v1 format to maximize compatibility

**Comment by munnerz** [comment]
> Would maybe want to define the normalized serialization so that a field selector that doesn't make use of in/notin with multiple values would still serialize in v1 format to maximize compatibility

I like the idea of this as it means existing output for users rendering field selectors won't change (assuming you mean serialising a 'basic' selector to the existing v1 format, therefore making the canonical way to specify exact equal semantics in 'v2' to be identical to what people use/see today).

**Comment by liggitt** [comment]
> assuming you mean serialising a 'basic' selector to the existing v1 format

yeah... I'd even consider saying the v2 format can *only* be used if you're using multi-item in/notin expressions

**Comment by jpbetz** [comment]
@deads2k @sttts and briefly chatted about this.  @deads2k brought up an interesting idea-- What if we used this as an opportunity to fix field selectors to actually match the exact paths to fields?  This could fix a problem we have today where we can't do client side field selection for built-in field selectors because there is [not a 1:1 mapping between field selector name and field paths for buil-in types](https://github.com/kubernetes/enhancements/blob/master/keps/sig-api-machinery/4358-custom-resource-field-selectors/README.md#background-field-selection-as-it-exists-before-this-enhancement).

If we introduced a `selector` query param (and did the 3 release soak to get it into all supported versions) and introduce a grammar where all field selectors AND all label selectors are specified using JSON path (.metadata.labels.labelname for label selectors) then we'd not only be able to unify how we handle field and label selectors AND also add the capability to process selectors client side.

**Comment by munnerz** [comment]
I think enforcing (correct) JSONPath syntax makes a tonne of sense. 

I think we may also still want to consider encoding a simple string that breaks older apiservers into the old field (iff the newer syntax is used) beyond the 3 version skew given how bad it'd be for a new-style field selector to be submitted to an older apiserver.

Would we still encode 'simple selectors' into the older fields for maximum compatibility of newer client libraries with older apiservers as Jordan suggested?

**Comment by munnerz** [comment]
Are there any known instances of field selectors that use valid json path keys, but return data that isn't sourced from that same field?

**Comment by munnerz** [comment]
I'd be happy to put together a KEP for this if we can come to a (rough!) agreement on the best approach. If we were to introduce a new 'Selector' field to ListOptions and DeleteOptions, would we take this as the opportunity/time to introduce the new 'selector' parameter (only) to other Options types like Create, Update and possibly also Get as discussed briefly in the selector based authZ KEP?

Do you foresee us introducing the new Selector type as a totally unified, technically field selector only syntax? (just acknowledging that .metadata.labels is a field already?). If so, would we encode 'label selectors' from the new selector field into the existing labelSelector query parameter for List/DeleteOptions? I think doing the same for field selectors could be tricky when handling the existing field selectors for non-existent paths.

Which leads me to think having usage of the new 'selector' field always require interaction with a compatible apiserver to use (by 'breaking' old selectors with the 'v2 trick') may be the most predictable approach for newer clients. The drawback is that if a client wants to specify a fieldSelector (say, for a non-existing field) *as well as* a new-style selector, they won't be able to without breaking the assurance that that sending the request to an out-of-support client<>server skew supported version is safe (which is quite scary given a breaking change of this impact/magnitude has not been done in the client for a long time/ever).

**Comment by jpbetz** [comment]
> I'd be happy to put together a KEP for this if we can come to a (rough!) agreement on the best approach.

Bringing a quick summary of the idea sig-api-machinery's semi-weekly meeting might be a good way to get the attention of a wider audience and get some early feedback before writing up the full KEP.

> Do you foresee us introducing the new Selector type as a totally unified, technically field selector only syntax? (just acknowledging that .metadata.labels is a field already?)

This approach appeals to me.  I haven't gotten my head around how to deal with the fact that label keys/values have a sufficiently constraint character set to make it safe to use bare strings, but that field selectors can be against arbitrary strings where bare strings seem far more problematic.  Figuring out what grammar to use feels like a big challenge with this idea.

I have limited time to think about this one until after 1.32 code freeze. But I would be great to have you drive this @munnerz

**Comment by sftim** [comment]
This issue also reminds me of https://github.com/kubernetes/kubernetes/issues/1348

**Comment by sftim** [comment]
Also see https://github.com/kubernetes/kubernetes/issues/1362

**Comment by munnerz** [comment]
I have begun working on a KEP to detail some of the use-cases here and am taking the preliminary approach of adding a new `selector` query parameter which accepts exclusively JSONPath syntax for selector keys.

I intend to include the details here around representing the 'canonical value' of 'simple' field selectors in their existing forms too, and likely use the `fieldSelector=v2` approach to 'break' older apiservers if we cannot represent the canonical form using the older fieldSelector and labelSelector syntax.

/assign

@sftim I think expanding selector syntax further may definitely have use-cases, and I think we should ensure that our chosen selector syntax should accommodate the addition of further operators too (once I have the KEP open, perhaps you could add a comment/some thoughts on how we can represent these things with the proposed selector format). I think logical OR is a more complex one/may require _additional_ query parameters to be added beyond `selector` (unless we want to start supporting `OR(AND(A, B), AND(C, D))` and other more complex chaining of selection criteria..).

> Features you could imagine but belong in a separate discussion:

Joe also suggested whether we could align the grammar to a restricted subset of CEL to allow us to potentially expand functionality even further in future. My concern with CEL in a user list/watch request is that we are effectively adding totally arbitrary filtering criteria onto the read path of all list/watches.

I think this is more closely aligned to the request for "arbitrary field selectors on core types" though, as we are basically permitting dynamic/on-demand indexing of object attributes. With field selectors today, the selectable-fields must be provided up-front (we don't index the entire object). For labels, the keys are available up front when the object is being persisted into the watch cache memory. With a dynamic CEL selector, we'd not know what fields to index until the time of the user request.

I'm not against the idea of aligning the syntax, but I wonder if we are better to have something like `?selector=JSONPath=.metadata.namespace IN (a, b, c)` vs `?selector=CEL=<cel-snippet>` instead, so we can avoid pushing a syntax we may never expand on in future on users?

**Comment by munnerz** [comment]
This would also give us time to bake apiserver support for parsing `?selector` into a 'type' and 'value' pair early on/sooner rather than later, which makes it far easier to introduce new selector formats/changes to the syntax without over-selecting against older apiservers too.

OTOH, this is where we *could* just re-use the `fieldSelector` field anyway, with a format like:

`?fieldSelector=jsonpath,<selector-value>`... `?fieldSelector=cel,<cel-value>` as we know that older apiservers already 'break' if any element in the selector slice (first split on `,`) does not contain an `=` sign...

**Comment by munnerz** [comment]
The more I think about this, the more I think re-using `?fieldSelector` makes sense. We could also then permit mixing together requirement syntaxes, e.g.:

```
?fieldSelector=spec.nodeName=X,jsonpath,.metadata.namespace IN (a, b, c),jsonpath,.metadata.name = Y
```

By reusing the existing field, we also prevent any accidents if a third-party client sets `?selector=x` *without* setting `?fieldSelector=dont-break-me-please` when talking to any apiserver that we have out there today (i.e. those that aren't aware of the `?selector` parameter yet).

I think this also leaves the door open for aggregated apiservers to support other types of field selection too, in case someone wants to experiment adding new syntaxes/expose querying features over their storage backend). I wonder how this extensibility mechanism could be used in something like metrics-server for example?

**Comment by jpbetz** [comment]
> By reusing the existing field, we also prevent any accidents if a third-party client sets `?selector=x` _without_ setting `?fieldSelector=dont-break-me-please` when talking to any apiserver that we have out there today (i.e. those that aren't aware of the `?selector` parameter yet).

So long as we don't have to trade away too much grammatically, I think I'd be okay with this. Moving this discussion to a KEP SGTM as well.

cc @deads2k

**Comment by thockin** [comment]
I only just became aware of this proposal. I support the expansion of semantics, but I want to be careful about defining a new, custom grammar.  We have a few places in the system where we have sort of ad hoc grammars with insufficient documentation and no aim towards consistency.  

Is it possible that this could be cel or maybe a subset? 

If not, then I think we should do a careful review of the grammar to make sure we handle what we need today, without cutting off tomorrow.

**Comment by sftim** [comment]
+1 for the subset-of-something approach

**Comment by mtaufen** [comment]
+1 to CEL if performance implications can be reasonably mitigated. 

As an aside, I'd like to see ALL requests support a consistent filtering semantic, not just aggregate requests. E.g. I'd like to be able to put selectors on a GET request and get a 404 if the target doesn't match (analogous to a resource being filtered out of a list request). And I'd like something similar for writes ("here's the selector for the thing(s) I'm trying to delete, don't let me delete anything that doesn't match").

**Comment by jpbetz** [comment]
> Is it possible that this could be cel or maybe a subset?

Yes, we could choose to use a small subset of CEL.  The CEL library provides a parser that outputs an AST.  We could elect to only allow a very small set of operators in CEL and effectively get a parser of a subset of the CEL grammar.  I'd very much like this to be considered amongst the alternatives of this proposal.

+100 to NOT introducing yet another completely bespoke grammar here.

**Comment by munnerz** [comment]
I think it makes sense to standardise on CEL, especially if we can use the AST to enforce rules on expressible selectors.

Do you think expressing 'or' semantics over multiple keys would also be in scope for this too? If so, we'd need to make specific optimisations in the watch cache to support this as it's a new 'type' of query rather than just a new grammar. (neither label or field selectors support OR semantics today). I'm inclined to push this to a later enhancement given its more unique/unknown performance implications.

**Comment by munnerz** [comment]
> As an aside, I'd like to see ALL requests support a consistent filtering semantic, not just aggregate requests. E.g. I'd like to be able to put selectors on a GET request and get a 404 if the target doesn't match (analogous to a resource being filtered out of a list request). And I'd like something similar for writes ("here's the selector for the thing(s) I'm trying to delete, don't let me delete anything that doesn't match"). 

+1 - this was also called out (and spec'd) in the [authorize with selectors KEP](https://github.com/kubernetes/enhancements/blob/master/keps/sig-auth/4601-authorize-with-selectors/README.md).

I'd expect any enhancement that adds selectors to other request verbs/types to also support a new CEL based selector format too.

**Comment by liggitt** [comment]
> Yes, we could choose to use a small subset of CEL. The CEL library provides a parser that outputs an AST.

That could make sense. I'd want to see us prove:
* it is efficient enough
* it can quickly reject syntax outside the allowed subset without fully parsing (so we don't let the CEL parser do expensive things on a giant expression we will just reject)
* we can fully constrain the syntax
* we can still transform it into our existing structured selector requirements: https://github.com/kubernetes/kubernetes/blob/bcd65ce24025e7d91b4c65be3fd3e38a2618bee5/staging/src/k8s.io/apimachinery/pkg/apis/meta/v1/types.go#L1326-L1341

>> As an aside, I'd like to see ALL requests support a consistent filtering semantic, not just aggregate requests. E.g. I'd like to be able to put selectors on a GET request and get a 404 if the target doesn't match (analogous to a resource being filtered out of a list request). And I'd like something similar for writes ("here's the selector for the thing(s) I'm trying to delete, don't let me delete anything that doesn't match").
>
> +1 - this was also called out (and spec'd) in the [authorize with selectors KEP](https://github.com/kubernetes/enhancements/blob/master/keps/sig-auth/4601-authorize-with-selectors/README.md).

We defined what it *would* mean *if* it was added. I'm not sure it's a good idea to add.

> Do you think expressing 'or' semantics over multiple keys would also be in scope for this too? 

I am not in favor of that. Adding the ability to express `in`/`notin` on a single key is the extent of what I'm looking for here.

**Comment by sftim** [comment]
[Maybe] we should move this to a Google doc as a precursor to a KEP. I think so anyway.
